tools/scanner/src/analysis.rs

```diff
@@ -461,50 +494,94 @@ impl FnLoops {
 struct IteratorVisitor<'a> {
     props: FnLoops,
     body: &'a Body,
+    visited_blocks: HashSet<usize>,
 }
 
 impl<'a> MirVisitor for IteratorVisitor<'a> {
+    fn visit_basic_block(&mut self, bb: &BasicBlock) {
+        self.visited_blocks.insert(self.body.blocks.iter().position(|b| *b == *bb).unwrap());
+        self.super_basic_block(bb);
+    }
+
     fn visit_terminator(&mut self, term: &Terminator, location: Location) {
-        if let TerminatorKind::Call { func, .. } = &term.kind {
-            let kind = func.ty(self.body.locals()).unwrap().kind();
-            if let TyKind::RigidTy(RigidTy::FnDef(def, _)) = kind {
-                let fullname = def.name();
-                let names = fullname.split("::").collect::<Vec<_>>();
-                if let [.., s_last, last] = names.as_slice() {
-                    if *s_last == "Iterator"
-                        && [
-                            "for_each",
-                            "collect",
-                            "advance_by",
-                            "all",
-                            "any",
-                            "partition",
-                            "partition_in_place",
-                            "fold",
-                            "try_fold",
-                            "spec_fold",
-                            "spec_try_fold",
-                            "try_for_each",
-                            "for_each",
-                            "try_reduce",
-                            "reduce",
-                            "find",
-                            "find_map",
-                            "try_find",
-                            "position",
-                            "rposition",
-                            "nth",
-                            "count",
-                            "last",
-                            "find",
-                        ]
-                        .contains(last)
-                    {
-                        self.props.iterators += 1;
+        match &term.kind {
```

Comment:
I think you can simplify this by calling `successors()` in the terminator before the match statement: https://github.com/rust-lang/rust/blob/4fe1e2bd5bf5a6f1cb245f161a5e9d315766f103/compiler/stable_mir/src/mir/body.rs#L130